switch to xtimer

kaspar030 opened the pull request:
This PR contains changes that drop hwtimer+vtimer and use periph/timer and xtimer instead.

~~Depends on #2926, those commits will go away as soon as xtimer itself is in.~~ (merged)

I've included a small (and inefficient) vtimer wrapper which makes using xtimer a drop-in replacement. Even using that wrapper, the code and memory requirements are lower than using the original hwtimer+vtimer code **in most cases**, with a lot of room for improvements.

~~This is work in progress. A lot else will not work at the moment. I'm putting it out so people can comment and collaborate early.~~ (no WIP anymore)

~~(Depends on #3837)~~

haukepetersen commented:
Just successfully tested with the new Silicon Labs SLWSTK6220A board - which uses a 16-bit timer.

OlegHahm commented:
> Just successfully tested with the new Silicon Labs SLWSTK6220A board - which uses a 16-bit timer.

Yay! :beers:

haukepetersen commented on cpu/native/include/periph_conf.h:
```diff
@@ -52,6 +52,9 @@
 /** @} */
 
+#define XTIMER_OVERHEAD 14
+#define XTIMER_USLEEP_UNTIL_OVERHEAD 1
```
how about spending these some doxygen...

haukepetersen commented on cpu/stm32f1/periph/timer.c:
```diff
@@ -94,7 +94,7 @@ int timer_init(tim_t dev, unsigned int ticks_per_us, void (*callback)(int))
     timer0->CR2 |= TIM_CR2_MMS_1;
     /* set auto-reload and prescaler values and load new values */
-    timer0->ARR = TIMER_0_MAX_VALUE;
+    timer0->ARR = 0xFFFF;
```
why? In the periph_conf you set `TIMER_0_MAX_VALUE` to 0xffffffff

kaspar030 replied:
As far as I understand the implementation, two 16bit timers are linked to one 32bit timer. So TIMER_0 has max_value=2^32, but the auto-reload register of one half can only take 16bit.

haukepetersen replied:
yapp, you're right. Should have looked at some more than just these few lines...

haukepetersen commented on sys/include/vtimer.h:
```diff
@@ -1,158 +1,29 @@
- * @{
- * @file
- * @author Kaspar Schleiser <kaspar@schleiser.de>
+ * @brief IPC message type for vtimer msg callback
```
please leave the license and doxygen in place.

haukepetersen commented on sys/include/xtimer.h:
```diff
@@ -0,0 +1,415 @@
+ * realized by next-first singly linked lists.
+ *
+ * @{
+ * @file
```
@brief missing.

kaspar030 replied:
you mean for the file?

haukepetersen replied:
yap

haukepetersen commented on sys/include/xtimer.h:
```diff
@@ -0,0 +1,415 @@
+ * interrupt context for the specified amount in *seconds*, so don't *ever* use
+ * it there,
+ *
+ * @param in seconds the amount of seconds the thread should sleep
```
I think doxygen wants `@param[in] seconds   ...` here.

haukepetersen commented:
only some minor stuff (@kaspar: I know how you just love doxygen :-) ). Works with various boards with different sized timers.

Regarding the backoff times: Can't we just define them once for every CPU family, and compute it somehow from the selected core clock value? This could save some configuration for each board/cpu...

kaspar030 commented:
> @kaspar: I know how you just love doxygen

IMHO as a programmer you've made it if you can just pay someone to deal with doxygen. ;)

kaspar030 commented:
> Regarding the backoff times: Can't we just define them once for every CPU family, and compute it somehow from the selected core clock value? This could save some configuration for each board/cpu...

Probably. The values are dependent on clock-speed, periph/timer implementation and a CPU's interrupt handling speed. I thought about calibrating the values at boot time, if we could spare a tenth of a second there, or, as a first step, writing a test that determines the correct values.

kaspar030 commented:
- addressed xtimer comments in parent PR (#2926)
- adressed other comments here

kaspar030 commented:
- rebased

haukepetersen commented:
could you please rebase?

kaspar030 commented:
- rebased

kaspar030 commented:
- rebased

kaspar030 commented:
@haukepetersen @PeterKietzmann Could you please take a look at 9df7ee3? The driver used microseconds for all delays but ```DELAY_CS_TOGGLE_TICKS``` and ```DELAY_AFTER_FUNC_TICKS```. Are 2us a sensible conversion?

kaspar030 commented:
- rebased

kaspar030 commented:
- rebased again

haukepetersen commented:
@kaspar030 the changes to the NRF radio driver are fine.

OlegHahm commented:
Is this still WIP?

kaspar030 commented:
@OlegHahm Yes, there are some uses of hwtimer left. working on it.

kaspar030 commented:
- rebased

OlegHahm commented:
btw `+615 −3,663` - nice!

OlegHahm commented:
Can you rebase, please?

kaspar030 commented:
- rebased

OlegHahm commented:
iotlab-m3 hangs sometimes after boot.
gdb shows:
```
xtimer_spin_until (value=28588) at /home/oleg/git/RIOT/sys/include/xtimer.h:424
424	    while (_xtimer_now() > value);
(gdb) bt
#0  xtimer_spin_until (value=28588) at /home/oleg/git/RIOT/sys/include/xtimer.h:424
#1  0x08008b54 in xtimer_set (timer=timer@entry=0x2000083c <main_stack+1344>, offset=offset@entry=1) at /home/oleg/git/RIOT/sys/xtimer/xtimer_core.c:125
#2  0x08008b80 in _xtimer_set64 (timer=timer@entry=0x2000083c <main_stack+1344>, offset=offset@entry=1, long_offset=long_offset@entry=0) at /home/oleg/git/RIOT/sys/xtimer/xtimer_core.c:92
#3  0x08008c4a in _xtimer_sleep (offset=offset@entry=1, long_offset=long_offset@entry=0) at /home/oleg/git/RIOT/sys/xtimer/xtimer.c:50
#4  0x08008d04 in xtimer_usleep (offset=1) at /home/oleg/git/RIOT/sys/include/xtimer.h:435
#5  at86rf2xx_reset (dev=dev@entry=0x20000d8c <at86rf2xx_devs>) at /home/oleg/git/RIOT/drivers/at86rf2xx/at86rf2xx.c:100
#6  0x08008e66 in at86rf2xx_init (dev=dev@entry=0x20000d8c <at86rf2xx_devs>, spi=spi@entry=SPI_0, spi_speed=spi_speed@entry=SPI_SPEED_5MHZ, cs_pin=cs_pin@entry=1073809412, int_pin=int_pin@entry=1073811460, sleep_pin=sleep_pin@entry=1073809410, 
    reset_pin=reset_pin@entry=1073811457) at /home/oleg/git/RIOT/drivers/at86rf2xx/at86rf2xx.c:85
#7  0x080012b2 in auto_init_at86rf2xx () at /home/oleg/git/RIOT/sys/auto_init/netif/auto_init_at86rf2xx.c:49
#8  0x08001286 in auto_init () at /home/oleg/git/RIOT/sys/auto_init/auto_init.c:148
#9  0x080008ba in main_trampoline (arg=<optimized out>) at /home/oleg/git/RIOT/core/kernel_init.c:54
#10 0x080002e4 in sched_switch (other_prio=36399) at /home/oleg/git/RIOT/core/sched.c:174
#11 0x20000824 in main_stack ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

kaspar030 commented:
@OlegHahm Could you try increasing RESET_DELAY in l39 of drivers/at86rf2xx/at86rf2xx.c to 2?

OlegHahm commented:
Tried - still hanging.

kaspar030 commented:
Setting to 100 works around the issue. Somehow in the iot-lab, spinning with very low values doesn't work...

kaspar030 commented:
- rebased

kaspar030 commented:
@BytesGalore Could you take a look at the fib changes?

OlegHahm commented:
@kaspar030, afaik @BytesGalore is out of town for another week, but I will take a look and test.

OlegHahm commented:
Everything seems to work at least as good as in master. Ready to squash.

kaspar030 commented:
- squashed

OlegHahm commented:
Do you have an overview what is still missing?

kaspar030 commented:
> Do you have an overview what is still missing?

For travis to be happy, x86 needs periph/timer (#3630). Then there's some discrepancy in the nrf24l01p (should be easy to fix).

Then we could theoretically merge, but probably every board needs to be checked for correct periph/timer configuration.

OlegHahm commented:
On the SAMR21 we face the same problem as for the iotlab-m3: hangs at at86rf2xx_reset:
```
(gdb) bt
#0  0x00005790 in timer_read (dev=dev@entry=1) at /home/oleg/git/RIOT/cpu/samd21/periph/timer.c:252
#1  0x00007614 in _xtimer_now () at /home/oleg/git/RIOT/sys/include/xtimer.h:392
#2  0x0000761e in xtimer_spin_until (value=305) at /home/oleg/git/RIOT/sys/include/xtimer.h:425
#3  0x000078fe in xtimer_set (timer=timer@entry=0x200007dc <main_stack+1328>, offset=offset@entry=1) at /home/oleg/git/RIOT/sys/xtimer/xtimer_core.c:125
#4  0x0000791e in _xtimer_set64 (timer=timer@entry=0x200007dc <main_stack+1328>, offset=offset@entry=1, long_offset=long_offset@entry=0) at /home/oleg/git/RIOT/sys/xtimer/xtimer_core.c:92
#5  0x000079f6 in _xtimer_sleep (offset=offset@entry=1, long_offset=long_offset@entry=0) at /home/oleg/git/RIOT/sys/xtimer/xtimer.c:50
#6  0x00007b10 in xtimer_usleep (offset=1) at /home/oleg/git/RIOT/sys/include/xtimer.h:436
#7  at86rf2xx_reset (dev=dev@entry=0x200012cc <at86rf2xx_devs>) at /home/oleg/git/RIOT/drivers/at86rf2xx/at86rf2xx.c:100
#8  0x00007c6c in at86rf2xx_init (dev=dev@entry=0x200012cc <at86rf2xx_devs>, spi=spi@entry=SPI_0, spi_speed=spi_speed@entry=SPI_SPEED_1MHZ, cs_pin=cs_pin@entry=1090536607, 
    int_pin=int_pin@entry=1090536576, sleep_pin=sleep_pin@entry=1090536468, reset_pin=reset_pin@entry=1090536591) at /home/oleg/git/RIOT/drivers/at86rf2xx/at86rf2xx.c:85
#9  0x00000db4 in auto_init_at86rf2xx () at /home/oleg/git/RIOT/sys/auto_init/netif/auto_init_at86rf2xx.c:49
#10 0x00000d92 in auto_init () at /home/oleg/git/RIOT/sys/auto_init/auto_init.c:148
#11 0x0000069a in main_trampoline (arg=<optimized out>) at /home/oleg/git/RIOT/core/kernel_init.c:54
#12 0x0000025c in sched_switch (other_prio=<optimized out>) at /home/oleg/git/RIOT/core/sched.c:180
#13 0x200007cc in main_stack ()
```
Increasing the `RESET_DELAY` to 10 fixes the issue.

OlegHahm commented:
But then it hangs occasionally here:
```
(gdb) bt
#0  timer_read (dev=dev@entry=1) at /home/oleg/git/RIOT/cpu/samd21/periph/timer.c:252
#1  0x00007614 in _xtimer_now () at /home/oleg/git/RIOT/sys/include/xtimer.h:392
#2  0x0000761e in xtimer_spin_until (value=40005039) at /home/oleg/git/RIOT/sys/include/xtimer.h:425
#3  0x000078fe in xtimer_set (timer=timer@entry=0x20001878 <ipv6_ifs+372>, offset=offset@entry=0) at /home/oleg/git/RIOT/sys/xtimer/xtimer_core.c:125
#4  0x0000791e in _xtimer_set64 (timer=timer@entry=0x20001878 <ipv6_ifs+372>, offset=0, long_offset=0) at /home/oleg/git/RIOT/sys/xtimer/xtimer_core.c:92
#5  0x00007a8a in xtimer_set_msg64 (timer=timer@entry=0x20001878 <ipv6_ifs+372>, offset=<optimized out>, msg=msg@entry=0x2000188c <ipv6_ifs+392>, target_pid=target_pid@entry=4)
    at /home/oleg/git/RIOT/sys/xtimer/xtimer.c:114
#6  0x000075fc in vtimer_set_msg (t=t@entry=0x20001878 <ipv6_ifs+372>, interval=..., pid=<optimized out>, type=type@entry=544, ptr=ptr@entry=0x20001704 <ipv6_ifs>)
    at /home/oleg/git/RIOT/sys/compat/vtimer/vtimer_compat.c:63
#7  0x00004f40 in _rtr_sol_reschedule (iface=iface@entry=0x20001704 <ipv6_ifs>, sec_delay=0) at /home/oleg/git/RIOT/sys/net/gnrc/network_layer/sixlowpan/nd/gnrc_sixlowpan_nd.c:33
#8  0x00005006 in gnrc_sixlowpan_nd_mc_rtr_sol (iface=0x20001704 <ipv6_ifs>) at /home/oleg/git/RIOT/sys/net/gnrc/network_layer/sixlowpan/nd/gnrc_sixlowpan_nd.c:85
#9  0x00001600 in _event_loop (args=<optimized out>) at /home/oleg/git/RIOT/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c:236
#10 0x0000025c in sched_switch (other_prio=<optimized out>) at /home/oleg/git/RIOT/core/sched.c:180
#11 0x77777776 in ?? ()
```

kaspar030 commented:
@OlegHahm Could you try https://gist.github.com/fad3c60b42298b0c86dc?

OlegHahm commented:
@kaspar030, thx. Seems to fix the problem indeed!

kaspar030 commented:
Travis is happy.

kaspar030 commented:
How do we proceed? Travis is happy, but this is probably the most intrusive change to RIOT, ever. Merge and write a note to devel@?

kaspar030 commented:
- rebased

kaspar030 commented:
double green! @OlegHahm already ACKed, would someone else?

cgundogan commented:
I tried simple networking with the `gnrc_networking`example and this PR merged into master. I couldn't see any strange behaviors that aren't in master already. So I guess I also ACK

OlegHahm commented:
@cgundogan, which platform?

cgundogan commented:
I tested on the iotlab testbed yesterday, briefly

OlegHahm commented:
`xtimer_usleep_until` fails on airfy-beacon:
```
2015-09-11 12:20:18,982 - INFO # Testing interval 12...
2015-09-11 12:20:18,985 - INFO # Testing interval 11...
2015-09-11 12:20:18,987 - INFO # Testing interval 10...
2015-09-11 12:20:18,989 - INFO # Testing interval 9...
2015-09-11 12:20:18,991 - INFO # Testing interval 8...
2015-09-11 12:20:18,993 - INFO # Testing interval 7...
2015-09-11 12:20:18,996 - INFO # Testing interval 6...
2015-09-11 12:20:35,775 - INFO # Testing interval 5...
```
Look at the last pyterm timestamps.

OlegHahm commented:
Tested successfully on
 * Arduino Due
 * iotlab-M3
 * mbed_lpc1768
 * nucleo-f091
 * nucleo-f334
 * pba-d-01-kw2x
 * stm32f0discovery

OlegHahm commented:
On samd21-xpro I observe similar behavior as on airfy-beacon.

OlegHahm commented:
Also on Zolertia Z1.

OlegHahm commented:
Tested successfully on native and avsextrem, too.

OlegHahm commented:
atmega2560 and nucleo-l1 has still some issues. As soon as these things are sorted out we're ready to merge this from my perspective.

sgso commented:
How is this supposed to work?
```C
    static inline void xtimer_spin_until(uint32_t value) {
        while (_xtimer_now() > value);
        while (_xtimer_now() < value);
    }
    /* in xtimer_usleep_until */
    uint32_t offset = target - now;
    if (offset > XTIMER_BACKOFF+XTIMER_USLEEP_UNTIL_OVERHEAD+1) {
       (...)
    }
    else {
        xtimer_spin_until(target);
    }
```
If `offset` is very small, `xtimer_spin_until` will loop until the timer overflows, at least that is what happens with my periph/timer implementation I'm building right now (and on samr21-xpro).

kaspar030 commented:
Alright, #3835 & #3837 contain fixes for atmega2560 and msp430.

OlegHahm commented:
With #3837 `xtimer_usleep_until` stopped working at all (at least on z1):
```
2015-09-14 16:14:54,693 - INFO # RIOT MSP430 hardware initialization complete.
2015-09-14 16:14:54,694 - INFO # 
2015-09-14 16:14:54,695 - INFO # main(): This is RIOT! (Version: 2014.12-3382-g95e3f-tbilisi)
2015-09-14 16:14:54,696 - INFO # xtimer_usleep_until test application.
2015-09-14 16:14:54,697 - INFO # 
2015-09-14 16:14:54,697 - INFO # Testing interval 1000...
```

OlegHahm commented:
The other xtimer tests seem to fail, too.

kaspar030 commented:
> The other xtimer tests seem to fail, too.

Could you try again with the latest commit, please?

OlegHahm commented:
Good news: the other timer tests run again.
Bad news: usleep_until still failing - but I guess that was expected.

kaspar030 commented:
I think I'll restrict periodic stuff (xtimer_sleep_until) to something like 1ms granularity. Having absolute targets with very low delays is hard to do right...

OlegHahm commented:
fine with me

kaspar030 commented:
- rebased
- improved xtimer_usleep_until()
- changed xtimer_xpin() so it cannot underflow for sane offsets

@OlegHahm Could you do another test round? 

xtimer_usleep_until() should finish now.
Some boards could be a little slow reading their timers, so if xtimer_usleep_until() doesn't finish, increasing xtimer's backoff time might help. (add ```CFLAGS += -DXTIMER_BACKOFF=40``` or even more to the makefile)

I tested arduino-mega2560, nucleo-f334, msb-430h, samr21-xpro and msba2.

kaspar030 commented:
> I think I'll restrict periodic stuff (xtimer_sleep_until) to something like 1ms granularity. Having absolute targets with very low delays is hard to do right...

I've changed the logic to use a double safety net.

OlegHahm commented on tests/xtimer_shift_on_compare/Makefile:
```diff
@@ -0,0 +1,12 @@
+APPLICATION = xtimer_shift_on_compare
+include ../Makefile.tests_common
+
+BOARD_WHITELIST += msb-430h arduino-mega2560
```
Why not whitelisting the other MSP430 boards, too?

OlegHahm commented:
Tested successfully on Z1, SAMR21-xpro, iotlab-m3, and avsextrem.

@haukepetersen, could you try with nucleo-l1 and airfy-beacon, please?

kaspar030 commented:
> nucleo-l1

I didn't continue fixing the l1. Not sure what's wrong there, but I 
strongly assume it's periph/timer.

kaspar030 commented:
@haukepetersen I didn't fix stmf4discovery, but maybe the general stability fixes make it work.

haukepetersen commented:
testing it just now

haukepetersen commented:
`airfy-beacon` and probably all other NRF51 based boards are broken -> I prepare a fix against your branch

haukepetersen commented:
done: https://github.com/kaspar030/RIOT/pull/14

haukepetersen commented:
The issue on the `stm32f4discovery` is still existent... Though the test applications all work fine - it only shows in the case as described in #3798. So I would say it is no a stopper for this PR!

kaspar030 commented:
- stm32f4discovery was too fast for the default backoff values. works now.

kaspar030 commented:
- removed dependency on  #3837 as this PR doesn't really depend on it

haukepetersen commented:
The problems with the `nucleo-l1` should be fixed with #3856

kaspar030 commented:
- rebased

haukepetersen commented:
ACK once Travis is green, second one needed...

@OlegHahm Now that the `nucleo-l1` is also fixed, is there any reason against merging this now (as soon as Travis is happy)?

kaspar030 commented:
- fixed and directly amended a license header mishap

OlegHahm commented:
I cannot test the l1 (USART is not running probably for weird reasons on my PC), so I cannot test, but I'm all in favor or merging this since all other boards I tested passed the tests and gnrc_networking runs also fine on iotlab-m3 and samr21-xpro.

This feels like the most heavyweight ACK ever: **ACK**

@kaspar030, awesome work! (Particular considering you reluctance to work ever on timers again.) This PR IMO improves RIOT a lot! I leave the honor of pushing the button to you - if you like.

haukepetersen commented:
Nice! So go Travis, go!

haukepetersen commented:
Travis failed due to some Travis internal things - but I am unable to restart those two jobs. Does somebody else have better luck?

kaspar030 commented:
No, can't restart either...

cgundogan commented:
me neither..

haukepetersen commented:
what do we do? I guess we do not expect build errors for those two builds, so merger anyway?

kaspar030 commented:
- null-rebased in order to get new commit so travis can do it's duty

haukepetersen commented:
ETA: 30min :-)

kaspar030 commented:
dammit, I got a flight to catch :)

haukepetersen commented:
Come on, the flight can't be more important then merging this PR :-)

kaspar030 commented:
Yeah, but it would be inconvenient to miss it. :)

kaspar030 commented:
I see green!

Everybody, thanks a lot for the painful reviewing! Let's cross fingers that this probably most intrusive change since RIOT's existence will turn out to be good. ;)

GO.

cgundogan commented:
Kudos!

emmanuelsearch commented:
awesome!!!

gebart commented:
Good (massive) work!

authmillenon commented:
Yay. gz.

thomaseichinger commented:
Congratulations, to all of you!
